AX: Dynamic page changes to iframes can cause their scroll-view to be set as the root, blocking assistive technologies from accessing the rest of the page content

twilco opened the pull request:
#### 6e9e17c1ad698a1bd5cb4989dc395f55df73f426
<pre>
AX: Dynamic page changes to iframes can cause their scroll-view to be set as the root, blocking assistive technologies from accessing the rest of the page content
<a href="https://bugs.webkit.org/show_bug.cgi?id=296298">https://bugs.webkit.org/show_bug.cgi?id=296298</a>
<a href="https://rdar.apple.com/156348748">rdar://156348748</a>

Reviewed by Joshua Hoffman.

Prior to <a href="https://commits.webkit.org/297651@main">https://commits.webkit.org/297651@main</a>, it was possible for an iframe&apos;s scroll-area child to be set as the
root of the isolated tree, preventing assistive technologies from accessing content outside of the iframe. This was
the sequence that caused this to happen:

  1. The iframe initially has a renderer, and is added to the isolated tree through normal means

  2. The iframe loses its renderer, e.g. via dynamic `display` change to `contents` or `none`

  3. Immediately after, something causes the scroll view child of the iframe to be queued up for a full isolated node
     update. One way this could happen is aria-hidden becoming false on the iframe, which would cause some or all descendants
     of the iframe to become unignored, which currently queues a full node update.

  4. Something initiates a flush of queued isolated tree updates, e.g. we are about to post a notification and need to
     flush queued updates to ensure we serve the request with an up-to-date tree.

  5. We run AXIsolatedTree::nodeChangeForObject. Because AccessibilityScrollView::parentObject used to return nullptr
     whenever its associated m_frameOwnerElement had no renderer, our scroll-view node change queued in step 3 thinks
     it has no parent. Prior to this commit, nodeChangeForObject used this as a signal that this was the root object,
     and set this iframe scroll-view as the root.

<a href="https://commits.webkit.org/297651@main">https://commits.webkit.org/297651@main</a> fixed this by making AccessibilityScrollView::parentObject more resilient. This
commit further strengthens our logic here by creating an explicit AccessibilityScrollView::isRoot() function, and using
that in nodeChangeForObject.

In order to test this, WKAccessibilityRootObject needed changes, as it always used the main-thread to get the root object,
meaning it wasn&apos;t ever actually testing the root that real clients see. This commit makes WKAccessibilityRootObject
threadsafe, and updates AccessibilityController::rootElement to call that function on the AX thread when available.
This commit also makes the fact that WKAccessibilityRootObject should only be used for testing explicit by renaming it
to _WKAccessibilityRootObjectForTesting. I did this because making this function threadsafe requires a synchronous main-thread
hit to use the AXObjectCache. Synchronous main-thread hits are OK for testing, but not for real clients, so this renaming
makes it clear that we don&apos;t need to try to move this usage off the main-thread.

With this new testing fixture in place, the added layout test (accessibility/mac/invalid-tree-root-at-iframe.html)
reproduced the bug 100% of the time.

This commit also cleans up a couple incorrect comments that confused me while debugging this issue (one in
AXObjectCache::handleRoleChanged and another in AXIsolatedTree::storeTree).

* LayoutTests/accessibility/mac/invalid-tree-root-at-iframe-expected.txt: Added.
* LayoutTests/accessibility/mac/invalid-tree-root-at-iframe.html: Added.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleRoleChanged):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isRoot const):
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::AccessibilityScrollView):
(WebCore::AccessibilityScrollView::isRoot const):
(WebCore::AccessibilityScrollView::ownerDebugDescription const):
* Source/WebCore/accessibility/AccessibilityScrollView.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::storeTree):
(WebCore::AXIsolatedTree::nodeChangeForObject):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFrame.cpp:
(_WKAccessibilityRootObjectForTesting):
(WKAccessibilityRootObject): Deleted.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFramePrivate.h:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityController.cpp:
(WTR::AccessibilityController::rootElement):
* Tools/WebKitTestRunner/InjectedBundle/atspi/AccessibilityControllerAtspi.cpp:
(WTR::AccessibilityController::accessibleElementById):
(WTR::AccessibilityController::rootElement):
(WTR::AccessibilityController::focusedElement):
* Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityControllerIOS.mm:
(WTR::AccessibilityController::focusedElement):
(WTR::AccessibilityController::accessibleElementById):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityControllerMac.mm:
(WTR::AccessibilityController::focusedElement):
(WTR::AccessibilityController::accessibleElementById):
(WTR::AccessibilityController::printTrees):

Canonical link: <a href="https://commits.webkit.org/297755@main">https://commits.webkit.org/297755@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ae2ea4a403e5f6cc6aed49527e61fb712152ff7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112706 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32438 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22916 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118905 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63201 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114668 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33090 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41001 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85781 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/36421 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115653 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26407 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101397 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66089 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25706 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19527 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62664 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95810 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19604 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122126 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39780 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29650 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94643 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40163 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97630 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94385 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24104 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39501 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17315 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35880 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39668 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45156 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39307 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42640 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41046 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AccessibilityScrollView.h:
```diff
@@ -85,11 +88,11 @@ class AccessibilityScrollView final : public AccessibilityObject {
     SingleThreadWeakPtr<ScrollView> m_scrollView;
     WeakPtr<HTMLFrameOwnerElement, WeakPtrImplWithEventTargetData> m_frameOwnerElement;
     RefPtr<AccessibilityObject> m_horizontalScrollbar;
     RefPtr<AccessibilityObject> m_verticalScrollbar;
-    bool m_childrenDirty;
```
Moving this field makes AccessibilityScrollView instances 8 bytes smaller by reducing unnecessary padding.

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityScrollView.cpp:
```diff
@@ -53,6 +53,28 @@ AccessibilityScrollView::~AccessibilityScrollView()
+String AccessibilityScrollView::ownerDebugDescription() const
+{
+    if (!m_frameOwnerElement)
+        return "null frame owner"_s;
```
Nit: could be nice to print whether this is the root or not (in the case of the comment you describe above, where `m_frameOwnerElement` was destroyed

twilco replied:
Fixed in latest commit!

hoffmanjoshua commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -211,9 +211,6 @@ void AXIsolatedTree::storeTree(AXObjectCache& cache, const Ref<AXIsolatedTree>&
 {
     ASSERT(isMainThread());
 
-    // Once we set this tree in the AXTreeStore, the secondary thread can start using it,
-    // and we can no longer access AXIsolatedTree::rootNode off the main-thread. Set the
-    // root now while we still can.
```
Rather than removing this outright, could we replace it with the correct explanation of what is happening here?

twilco replied:
Fixed in latest commit!

hoffmanjoshua commented on Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFrame.cpp:
```diff
@@ -38,6 +38,7 @@
 #include "WKData.h"
 #include "WebFrame.h"
 #include "WebPage.h"
+#include <WebCore/AXIsolatedObject.h>
```
Do you recall why this import is needed? Is it for `AXIsolatedTree`?

twilco replied:
It's necessary to make the below `RefPtr`, since `AXIsolatedTree::rootNode` returns an `AXIsolatedObject*`, and constructing a RefPtr of something requires the header to be imported.

`RefPtr root = tree ? tree->rootNode() : nullptr;`

hoffmanjoshua reviewed (approved):
Everything looks good to me! Just a few small questions/nits

minorninth commented on LayoutTests/accessibility/mac/invalid-tree-root-at-iframe.html:
```diff
@@ -0,0 +1,77 @@
+</head>
+<body>
+
+<iframe id="iframe" onload="runTest()" aria-hidden="true" srcdoc="<html aria-label='iframe-web-area'><button>Inside iframe</button></html>"></iframe>
```
I think you should be able to put newlines in srcdoc, which might make this more readable. (Unless there's a reason that breaks the test.)

twilco replied:
Done in latest commit!

minorninth commented on Source/WebCore/accessibility/AccessibilityObject.h:
```diff
@@ -123,6 +123,7 @@ class AccessibilityObject : public AXCoreObject {
     bool isMediaTimeline() const { return false; }
     virtual bool isSliderThumb() const { return false; }
     bool isLabel() const { return isAccessibilityLabelInstance() || labelForObjects().size(); }
+    virtual bool isRoot() const { return false; }
```
Maybe this should have a comment. We will have to figure out what this means in the context of site isolation.

twilco replied:
Done in latest commit!

minorninth commented on Source/WebCore/accessibility/AccessibilityScrollView.cpp:
```diff
@@ -53,6 +53,33 @@ AccessibilityScrollView::~AccessibilityScrollView()
+    // doesn't guarantee that this is the root, as the element may have been destroyed but |this|
+    // is still valid and undestroyed for some amount of time.
+
+    return m_scrollView && !m_scrollView->parent();
```
If we're on the main thread, I feel like we should be able to get a 100% correct answer here, rather than using a heuristic that has a chance of being wrong under other weird circumstances.

One way: m_scrollView is a LocalFrameView, which has a LocalFrame, which has a localMainFrame - if the localFrame is not the same as the main frame then this AccessibilityScrollView is not the root.

Or alternatively if we cache AXObjectCache when constructing AccessibilityScrollView, then use the Document from AXObjectCache and go to Document > Frame > FrameView and see if it's the same as m_scrollView - if not, it's not the root.

twilco replied:
Done in latest commit!

minorninth commented on Tools/WebKitTestRunner/InjectedBundle/atspi/AccessibilityControllerAtspi.cpp:
```diff
@@ -63,7 +63,7 @@ static WebCore::AccessibilityObjectAtspi* findAccessibleObjectById(WebCore::Acce
 RefPtr<AccessibilityUIElement> AccessibilityController::accessibleElementById(JSContextRef context, JSStringRef id)
 {
-    auto* rootObject = static_cast<WebCore::AccessibilityObjectAtspi*>(WKAccessibilityRootObject(WKBundleFrameForJavaScriptContext(context)));
+    auto* rootObject = static_cast<WebCore::AccessibilityObjectAtspi*>(_WKAccessibilityRootObjectForTesting(WKBundleFrameForJavaScriptContext(context)));
```
Does everything called from a layout test with isolated-tree on go through _WKAccessibilityRootObjectForTesting now?

twilco replied:
Yes it does!

minorninth reviewed (approved):
Fantastic job on the regression test!

twilco committed:
AX: Dynamic page changes to iframes can cause their scroll-view to be set as the root, blocking assistive technologies from accessing the rest of the page content
https://bugs.webkit.org/show_bug.cgi?id=296298
rdar://156348748

Reviewed by Joshua Hoffman.

Prior to https://commits.webkit.org/297651@main, it was possible for an iframe's scroll-area child to be set as the
root of the isolated tree, preventing assistive technologies from accessing content outside of the iframe. This was
the sequence that caused this to happen:

  1. The iframe initially has a renderer, and is added to the isolated tree through normal means

  2. The iframe loses its renderer, e.g. via dynamic `display` change to `contents` or `none`

  3. Immediately after, something causes the scroll view child of the iframe to be queued up for a full isolated node
     update. One way this could happen is aria-hidden becoming false on the iframe, which would cause some or all descendants
     of the iframe to become unignored, which currently queues a full node update.

  4. Something initiates a flush of queued isolated tree updates, e.g. we are about to post a notification and need to
     flush queued updates to ensure we serve the request with an up-to-date tree.

  5. We run AXIsolatedTree::nodeChangeForObject. Because AccessibilityScrollView::parentObject used to return nullptr
     whenever its associated m_frameOwnerElement had no renderer, our scroll-view node change queued in step 3 thinks
     it has no parent. Prior to this commit, nodeChangeForObject used this as a signal that this was the root object,
     and set this iframe scroll-view as the root.

https://commits.webkit.org/297651@main fixed this by making AccessibilityScrollView::parentObject more resilient. This
commit further strengthens our logic here by creating an explicit AccessibilityScrollView::isRoot() function, and using
that in nodeChangeForObject.

In order to test this, WKAccessibilityRootObject needed changes, as it always used the main-thread to get the root object,
meaning it wasn't ever actually testing the root that real clients see. This commit makes WKAccessibilityRootObject
threadsafe, and updates AccessibilityController::rootElement to call that function on the AX thread when available.
This commit also makes the fact that WKAccessibilityRootObject should only be used for testing explicit by renaming it
to _WKAccessibilityRootObjectForTesting. I did this because making this function threadsafe requires a synchronous main-thread
hit to use the AXObjectCache. Synchronous main-thread hits are OK for testing, but not for real clients, so this renaming
makes it clear that we don't need to try to move this usage off the main-thread.

With this new testing fixture in place, the added layout test (accessibility/mac/invalid-tree-root-at-iframe.html)
reproduced the bug 100% of the time.

This commit also cleans up a couple incorrect comments that confused me while debugging this issue (one in
AXObjectCache::handleRoleChanged and another in AXIsolatedTree::storeTree).

* LayoutTests/accessibility/mac/invalid-tree-root-at-iframe-expected.txt: Added.
* LayoutTests/accessibility/mac/invalid-tree-root-at-iframe.html: Added.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleRoleChanged):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isRoot const):
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::AccessibilityScrollView):
(WebCore::AccessibilityScrollView::isRoot const):
(WebCore::AccessibilityScrollView::ownerDebugDescription const):
* Source/WebCore/accessibility/AccessibilityScrollView.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::storeTree):
(WebCore::AXIsolatedTree::nodeChangeForObject):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFrame.cpp:
(_WKAccessibilityRootObjectForTesting):
(WKAccessibilityRootObject): Deleted.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleFramePrivate.h:
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityController.cpp:
(WTR::AccessibilityController::rootElement):
* Tools/WebKitTestRunner/InjectedBundle/atspi/AccessibilityControllerAtspi.cpp:
(WTR::AccessibilityController::accessibleElementById):
(WTR::AccessibilityController::rootElement):
(WTR::AccessibilityController::focusedElement):
* Tools/WebKitTestRunner/InjectedBundle/ios/AccessibilityControllerIOS.mm:
(WTR::AccessibilityController::focusedElement):
(WTR::AccessibilityController::accessibleElementById):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityControllerMac.mm:
(WTR::AccessibilityController::focusedElement):
(WTR::AccessibilityController::accessibleElementById):
(WTR::AccessibilityController::printTrees):

Canonical link: https://commits.webkit.org/297755@main
